File: dataModule/src/main/java/com/mobnetic/coinguardian/model/market/Coinnest.java
package com.mobnetic.coinguardian.model.market;

import com.mobnetic.coinguardian.model.CheckerInfo;
import com.mobnetic.coinguardian.model.Market;
import com.mobnetic.coinguardian.model.Ticker;
import com.mobnetic.coinguardian.model.currency.Currency;
import com.mobnetic.coinguardian.model.currency.VirtualCurrency;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.LinkedHashMap;

public class Coinnest extends Market {

    private final static String NAME = "Coinnest";
    private final static String TTS_NAME = NAME;
    private final static String URL_BASE = "https://api.coinnest.co.kr";
    private final static String URL_TICKER = URL_BASE + "/api/pub/ticker?coin=%1$s";
    private final static String URL_ORDERS = URL_BASE + "/api/pub/depth?coin=%1$s";
    private final static HashMap<String, CharSequence[]> CURRENCY_PAIRS = new LinkedHashMap<>();

    static {
        CURRENCY_PAIRS.put(VirtualCurrency.BTC, new String[]{
                Currency.KRW
        });
        CURRENCY_PAIRS.put(VirtualCurrency.BCC, new String[]{
                Currency.KRW
        });
        CURRENCY_PAIRS.put(VirtualCurrency.ETH, new String[]{
                Currency.KRW
        });
        CURRENCY_PAIRS.put(VirtualCurrency.ETC, new String[]{
                Currency.KRW
        });
        CURRENCY_PAIRS.put(VirtualCurrency.QTUM, new String[]{
                Currency.KRW
        });
        CURRENCY_PAIRS.put(VirtualCurrency.NEO, new String[]{
                Currency.KRW
        });
        CURRENCY_PAIRS.put(VirtualCurrency.KNC, new String[]{
                Currency.KRW
        });
        CURRENCY_PAIRS.put(VirtualCurrency.TSL, new String[]{
                Currency.KRW
        });
        CURRENCY_PAIRS.put(VirtualCurrency.OMG, new String[]{
                Currency.KRW
        });
    }

    public Coinnest() {
        super(NAME, TTS_NAME, CURRENCY_PAIRS);
    }

    @Override
    public int getNumOfRequests(CheckerInfo checkerRecord) {
        return 1;
    }

    @Override
    public String getUrl(int requestId, CheckerInfo checkerInfo) {
        return String.format(URL_TICKER, checkerInfo.getCurrencyBaseLowerCase());
    }

    @Override
    protected void parseTickerFromJsonObject(int requestId, JSONObject jsonObject, Ticker ticker,
                                             CheckerInfo checkerInfo) throws Exception {
        ticker.high = jsonObject.getDouble("high");
        ticker.low = jsonObject.getDouble("low");
        ticker.bid = jsonObject.getDouble("buy");
        ticker.ask = jsonObject.getDouble("sell");
        ticker.last = jsonObject.getDouble("last");
        ticker.vol = jsonObject.getDouble("vol");
        ticker.timestamp = jsonObject.getLong("time");
    }

    @Override
    protected String parseErrorFromJsonObject(int requestId, JSONObject jsonObject,
                                              CheckerInfo checkerInfo) throws Exception {
        return jsonObject.getString("msg");
    }
}

File: dataModule/src/main/java/com/aneonex/bitcoinchecker/datamodule/model/market/Coinnest.kt
package com.aneonex.bitcoinchecker.datamodule.model.market

import com.aneonex.bitcoinchecker.datamodule.model.CheckerInfo
import com.aneonex.bitcoinchecker.datamodule.model.Market
import com.aneonex.bitcoinchecker.datamodule.model.Ticker
import com.aneonex.bitcoinchecker.datamodule.model.currency.Currency
import com.aneonex.bitcoinchecker.datamodule.model.currency.CurrencyPairsMap
import com.aneonex.bitcoinchecker.datamodule.model.currency.VirtualCurrency
import com.aneonex.bitcoinchecker.datamodule.util.TimeUtils
import org.json.JSONObject

class Coinnest : Market(NAME, TTS_NAME, CURRENCY_PAIRS) {
    companion object {
        private const val NAME = "Coinnest"
        private const val TTS_NAME = NAME
        private const val URL = "https://api.coinnest.co.kr/api/pub/ticker?coin=%1\$s"
        private val CURRENCY_PAIRS: CurrencyPairsMap = CurrencyPairsMap()

        init {
            CURRENCY_PAIRS[VirtualCurrency.BTC] = arrayOf(
                    Currency.KRW
            )
            CURRENCY_PAIRS[VirtualCurrency.BCC] = arrayOf(
                    Currency.KRW
            )
            CURRENCY_PAIRS[VirtualCurrency.ETH] = arrayOf(
                    Currency.KRW
            )
            CURRENCY_PAIRS[VirtualCurrency.ETC] = arrayOf(
                    Currency.KRW
            )
            CURRENCY_PAIRS[VirtualCurrency.QTUM] = arrayOf(
                    Currency.KRW
            )
            CURRENCY_PAIRS[VirtualCurrency.NEO] = arrayOf(
                    Currency.KRW
            )
            CURRENCY_PAIRS[VirtualCurrency.KNC] = arrayOf(
                    Currency.KRW
            )
            CURRENCY_PAIRS[VirtualCurrency.TSL] = arrayOf(
                    Currency.KRW
            )
            CURRENCY_PAIRS[VirtualCurrency.OMG] = arrayOf(
                    Currency.KRW
            )
        }
    }

    override fun getUrl(requestId: Int, checkerInfo: CheckerInfo): String {
        return String.format(URL, checkerInfo.currencyBaseLowerCase)
    }

    @Throws(Exception::class)
    override fun parseTickerFromJsonObject(requestId: Int, jsonObject: JSONObject, ticker: Ticker,
                                           checkerInfo: CheckerInfo) {
        ticker.high = jsonObject.getDouble("high")
        ticker.low = jsonObject.getDouble("low")
        ticker.bid = jsonObject.getDouble("buy")
        ticker.ask = jsonObject.getDouble("sell")
        ticker.last = jsonObject.getDouble("last")
        ticker.vol = jsonObject.getDouble("vol")
        ticker.timestamp = jsonObject.getLong("time") * TimeUtils.MILLIS_IN_SECOND
    }

    @Throws(Exception::class)
    override fun parseErrorFromJsonObject(requestId: Int, jsonObject: JSONObject,
                                          checkerInfo: CheckerInfo?): String? {
        return jsonObject.getString("msg")
    }
}